sig-multicluster/namespace-sameness-position-statement.md

```diff
@@ -92,10 +96,29 @@ thus namespce sameness applies), the implementation of multi-cluster services
 probably should not be on-by-default.  The details of how multi-cluster
 services will work are an area for innovation, but ideas include:
   * Opt-in: services must be "exported" to be merged across clusters
-  * Opt-out: services or namespace can be opted out of service merging
+  * Opt-out: services or namespaces can be opted out of sameness
   * Different discovery: merged services and "raw" services use different names
     or other discovery mechanisms
 
 However the implementation works, the metrics service is not automatically
 merged across clusters, though the LDAP-to-RBAC sync from example 2 can still
 apply consistent policies.
+
+### Example 4: Authority sameness policy
+
+Consider the same organization from previous examples, but with more clusters:
+A, B, C, and D.  They want to assign foo-team's namespaces to clusters A and B,
+and bar-team's namespaces to clusters C and D.  They also want to ensure that
+the opposite is never true - foo-team's namespaces cannot be used in clusters C
+or D, and bar-team's namespaces cannot be used from clusters A or B.
+
+As an implementation choice, the authority may offers controls which govern
+which cluster-namespaces are considered for sameness.  For example, it could
+express "all namespaces prefixed with 'foo-' may be merged from clusters A and
+B" and "all namespaces prefixed with 'bar-' may be merged from clusters C and
+D".  If a "foo-something" namespace is found in cluster C it would _not_ be
+considered for sameness.  The details of how these policies might be expressed
```

Comment:
This doesn't feel workable for an MCS API implementation.

If _no_ Services in a `foo-something` namespace are intended to be exported for consumption by Services in `bar-something` namespaces, then these sets of clusters are fully independent and are _not_ "expected to work together".

If at least one Service in a `foo-something` namespace _is_ intended to be consumed from Services managed by `bar-team`, then:
  - Preventing creation of the `foo-something` namespace prevents placing any `ServiceImport` in it to route to a Service in `foo-something` on clusters A or B managed by `foo-team` - services wouldn't be able to be shared between teams.
  - Allowing a `foo-something` namespace to exist in cluster C but excluding all `foo-*` prefixed namespaces from "sameness" by policy would have one of the following effects:
    - Prevent the creation of a ServiceImport because the namespace is not the same - this isn't really practical because it prevents any service network between the teams.
    - Allow the creation of _only_ ServiceImport resources in the `foo-something` namespace - this effectively makes all `foo-*` namespaces on clusters C and D only usable for routing to `foo-team` services on clusters A or B - this is kinda the best case scenario, but has the downside of requiring global namespace management and complex RBAC/sameness policy.
    - Allow the creation of `ServiceImport` _and_ cluster-local `Service` or other resources, but _not_ consider a `ServiceImport` and `Service` with the same name in the `foo-something` namespace to logically identical or fungible - this could lead to significant user confusion.

Reply:
I'm not 100% sure I caught all that you are saying here, so correct me if I miss the point, please.  Note, my next push changes the scenario a bit WRT A, B, C, D and foo, bar, but I'll answer here with the scenario as written :)

Clusters can be "expected to work together" even if some namespaces in them are "private" to the cluster.  As an example, consider kube-system.  Every cluster has one, but cross-cluster "sameness" is dubious.  If you consider a clusterset as a management domain, and you buy the argument that multi-cluster and cross-cluster capabilities will continue to grow, then there's clearly value in having a smaller number of domains, even if any given capability may or may not apply to a given cluster.  IOW, some clusters will use MCS.  Some will use MC-Ingress.  Some will be CD targets.  Some will be build clusters.  Some will be dedicated.  Some will use gitops.  I don't think we can or should demand homegeneity here. 

> Preventing creation of the foo-something namespace prevents placing any ServiceImport in it

Correct.  I'm trying to thread a needle and allow the implementation of MCS to (for example) populate a ServiceImport in C/foo-something (with endpoints from A and B) but ALSO to ignore any ServiceExport it might find in C or D.  In effect, clusters C and D can *consume* foo-something services but not *produce* them.

> Allow the creation of only ServiceImport resources in the foo-something namespace - this effectively makes all foo-* namespaces on clusters C and D only usable for routing to foo-team services on clusters A or B - this is kinda the best case scenario, but has the downside of requiring global namespace management and complex RBAC/sameness policy.

I don't see the global namespace management as a "problem" really.  Or rather, it's a problem we already have, so it's not worse in this case.

> Allow the creation of ServiceImport and cluster-local Service or other resources, but not consider a ServiceImport and Service with the same name in the foo-something namespace to logically identical or fungible

I think the assertion is closer to your 2nd point ("the best case scenario") - there SHOULD NOT be foo-something cluster-local Services or other resources in cluster C.  But on the off chance that we find them (e.g. cluster was compromised or somehow fat-fingered) we will NOT engage them because the policy says they should not be there.

As I said somewhere else -  what I am trying to do here is allow implementations the freedom to explore.  I think different situations will need different options, and I found the language in here to be overly rigid.  I am not saying implementations MUST do any of this - just that they MAY.